### PR TITLE
Add companies postcode filter

### PR DESCRIPTION
**Pull request by elcct**
### Description of change

`POST /v4/search/company`: A `uk_postcode` filter was added for the `address_postcode` and `registered_address_postcode` 
fields for UK based companies. The filter accepts a single or a partial postcode as well as an array of postcodes. Multiple postcodes are 
matched with `or` query.

Because we only want to analyse UK postcodes I created two properties - `uk_address_postcode` and `uk_registered_address_postcode` that only return value of corresponding postcode fields if given company is UK based. These properties are indexed and analysed with the analysers described below.

`Elasticsearch`: The search of UK postcodes has been improved so that any entity having a field storing UK postcode 
should use new `postcode_analyzer` and `postcode_search_analyzer` analyzers. It is now possible to perform partial 
search for area, area and district, area and sub-district and also area, district and sector and whole postcode. 
Postcodes are case insensitive and any spaces are filtered out before analysis.

### Checklist

* [x] Has a new newsfragment been created? Check [changelog/README.md](https://github.com/uktrade/data-hub-api/blob/master/changelog/README.md) for instructions
* [ ] Do any added or updated endpoints appear in the API documentation? See [docs/Maintaining the API documentation.md](https://github.com/uktrade/data-hub-api/blob/develop/docs/Maintaining&#32;the&#32;API&#32;documentation.md) for more details
* [x] Have any relevant search models been updated?
* [ ] Have any relevant fixtures (`fixtures/test_data.yaml`) been updated?
* [ ] Have any relevant select-/prefetch-related field lists in the views and search apps been updated?
* [ ] Has the admin site been updated (for new models, fields etc.)?
* [ ] Has the README been updated (if needed)?

### CODE REVIEW

**Comment by reupen** [comment]
@elcct One of the requirements was that the filter only matches UK companies (and not foreign ones). Is that ensured here?

**Comment by elcct** [comment]
@reupen I missed that. I have modified it so it now uses separate fields for UK company postcodes. Thank you!

File: datahub/search/elasticsearch.py [hunk 1]
```diff
@@ -34,6 +34,59 @@
...
+        # Area and district
+        # Area + one or two numbers or a number and letter
+        '^([a-z]{1,2}(?:[0-9][a-z]|[0-9]{1,2}))[0-9][a-z]{2}',
+
+        # Area, district and sector
+        # Area + district + digit
+        '^([a-z]{1,2}(?:[0-9][a-z]|[0-9]{1,2})[0-9])[a-z]{2}',
+
+        # Whole postcode
+        '^([a-z]{1,2}(?:[0-9][a-z]|[0-9]{1,2})[0-9][a-z]{2})',
```
Comment (reupen) [inline datahub/search/elasticsearch.py]: What are the advantages of trying to parse the entire postcode vs leaving `preserve_original` as `True` (the default)? Just thinking it might be better to be forgiving for the entire postcode match.

Comment (elcct) [reply datahub/search/elasticsearch.py]: There are no advantages. I have removed the last pattern now and replaced it with `preserve_original`. Thank you!

File: datahub/company/models/company.py [hunk 2]
```diff
@@ -302,6 +302,20 @@ def uk_based(self):
...
+        if self.uk_based:
+            return self.address_postcode
+        return ''
+
+    @property
+    def uk_registered_address_postcode(self):
+        """Return `registered_address_postcode` if company is uk based."""
+        if self.uk_based:
+            return self.registered_address_postcode
+        return ''
```
Comment (reupen) [inline datahub/company/models/company.py]: Are we able to use `COMPUTED_MAPPINGS` on the search model instead of adding properties here as these seem very specific to the search fields?

**Comment by reupen** [comment]
Noticed that the new fields are appearing in responses – perhaps we could add them to `fields_to_exclude` on the view to prevent that.

File: datahub/search/fields.py [hunk 3]
```diff
@@ -17,6 +17,12 @@
         'trigram': TrigramText(),
     },
 )
+# Keyword with normalisation that recognises UK postcodes to improve searching
+PostcodeKeyword = partial(
+    Text,
+    analyzer='postcode_analyzer',
+    search_analyzer='postcode_search_analyzer',
+)
```
Comment (reupen) [inline datahub/search/fields.py]: We should be able to import the analysers and pass them in directly i.e.:

```python
PostcodeKeyword = partial(
    Text,
    analyzer=postcode_analyzer,
    search_analyzer=postcode_search_analyzer,
)
```

(That way, we don't need to add them to `ANALYZERS` in `datahub.search.elasticsearch`.)

File: datahub/search/elasticsearch.py [hunk 4]
```diff
@@ -34,6 +34,57 @@
...
+
+postcode_filter = analysis.token_filter(
+    'postcode_filter',
+    type='pattern_capture',
+    # Match whole postcode
+    preserve_original=True,
+    patterns=[
+        # Area
+        # One or two letters
+        '^([a-z]{1,2})(?:[0-9][a-z]|[0-9]{1,2})[0-9][a-z]{2}',
```
Comment (reupen) [inline datahub/search/elasticsearch.py]: Should we make these raw strings since they are regexes?

Also, I think it would be good to define them in named constants to rely on comments a bit less e.g.

```python
# First one or two letters
POSTCODE_AREA_REGEX = r'...'
```

File: datahub/search/elasticsearch.py [hunk 5]
```diff
@@ -34,6 +34,57 @@
...
+
+        # Area and district (with sub-district code ignored)
```
Comment (reupen) [inline datahub/search/elasticsearch.py]: Very minor thing but I think saying 'area and district' is a bit redundant (from [the PAF documentation](https://www.poweredbypaf.com/wp-content/uploads/2017/07/Latest-Programmers_guide_Edition-7-Version-6.pdf)):

image

(Would be good to add a comment saying something like 'See the Royal Mail programmer's guide for the exact definitions too.)

File: datahub/search/company/test/test_entity_search_views.py [hunk 6]
```diff
@@ -355,6 +399,30 @@ def test_response_body(self, es_with_collector):
...
+                [],
+            ),
+
+            # uk_postcode
+            (
+                {
+                    'uk_postcode': ['SW1', 'wc'],
+                },
+                ['abc defg ltd'],
+            ),
```
Comment (reupen) [inline datahub/search/company/test/test_entity_search_views.py]: Is there a reason for testing the filter here and in `test_search_postcodes` below? I think it would be good to do the testing in one place. 

(Also would be good to add a case with a space in the search term if we don't have one already.)

File: datahub/search/company/test/test_entity_search_views.py [hunk 7]
```diff
@@ -372,12 +440,81 @@ def test_filters(self, setup_data, filters, expected_companies):
...
+            ('W1', ['company_w1', 'company_w1a']),
+            ('W11', ['company_w11']),
+            ('W1A', ['company_w1a']),
+            ('W', ['company_w1', 'company_w1a', 'company_w11']),
+            ('WC2', ['company_wc2b', 'company_wc2n']),
+            ('WC', ['company_wc2b', 'company_wc2n', 'company_wc1x', 'company_wc1a']),
+            ('AB1', ['company_ab1']),
+            ('AB10', ['company_ab10']),
+            ('AB', ['company_ab1', 'company_ab10']),
+            ('SE1', ['company_se1', 'company_se1_2']),
```
Comment (reupen) [inline datahub/search/company/test/test_entity_search_views.py]: Perhaps these could be organised by whether an area, district, sub-district or sector is being searched for? Would make it a bit easier to see what's covered.

File: datahub/search/company/test/test_entity_search_views.py [hunk 8]
```diff
@@ -372,12 +440,81 @@ def test_filters(self, setup_data, filters, expected_companies):
...
+                    'company_w1',
+                    'company_w1a',
+                    'company_w11',
+                    'company_ab1',
+                    'company_ab10',
+                ],
+            ),
+
+            # Valid and invalid
+            (['SE1', 'Invalid'], ['company_se1', 'company_se1_2']),
```
Comment (reupen) [inline datahub/search/company/test/test_entity_search_views.py]: I'm not sure there is much benefit to having so many array cases since the OR logic should be being tested elsewhere already. I think one or two arrays cases should be sufficient.

File: datahub/search/elasticsearch.py (hunk 5 above)
Comment (elcct) [reply datahub/search/elasticsearch.py]: I see your point, but I was commenting about individual parts that whole postcode is composed of  that is area code, district, sector and unit and which ones will go into group to make a token.
If think the table above doesn't explain well what makes e.g. Postcode sector (that is area + district + sector). I am happy to modify the comments though and link to Royal Mail documentation.

Comment (elcct) [reply datahub/search/elasticsearch.py]: I have updated the comments now, I think together with the constants it looks a little better. Thank you!

File: datahub/search/company/test/test_entity_search_views.py (hunk 6 above)
Comment (elcct) [reply datahub/search/company/test/test_entity_search_views.py]: I added it because there are also other filters being tested, but since it also has a separate test, it will make sense to remove it from that test. Thank you!

File: datahub/search/company/test/test_entity_search_views.py (hunk 8 above)
Comment (elcct) [reply datahub/search/company/test/test_entity_search_views.py]: I agree. I removed some tests cases. Thank you!

File: datahub/search/company/test/test_entity_search_views.py (hunk 7 above)
Comment (elcct) [reply datahub/search/company/test/test_entity_search_views.py]: This is a good idea. I have organised the tests now. Thank you!

File: datahub/search/elasticsearch.py (hunk 4 above)
Comment (elcct) [reply datahub/search/elasticsearch.py]: I started doing this and realised maybe it is not as readable as I thought it will be, however, I see constants could help understand the patterns quicker, so I added it now. Thank you!

File: datahub/search/fields.py (hunk 3 above)
Comment (elcct) [reply datahub/search/fields.py]: I fixed that now. Thank you!

File: datahub/company/models/company.py (hunk 2 above)
Comment (elcct) [reply datahub/company/models/company.py]: That makes sense. There is no point to have these properties on the model. I have updated the search model now. Thank you!

File: datahub/search/elasticsearch.py [hunk 9]
```diff
@@ -41,30 +41,36 @@
...
     # Match whole postcode
     preserve_original=True,
     patterns=[
-        # Area
-        # One or two letters
-        '^([a-z]{1,2})(?:[0-9][a-z]|[0-9]{1,2})[0-9][a-z]{2}',
+        # Match postcode area
+        # See the Royal Mail programmer's guide for the exact definitions
+        f'^({POSTCODE_AREA_REGEX})(?:{POSTCODE_DISTRICT_REGEX})'
+        f'{POSTCODE_SECTOR_REGEX}{POSTCODE_UNIT_REGEX}',
```
Comment (reupen) [inline datahub/search/elasticsearch.py]: Should be able to make this raw f-strings as well (by using an `rf` prefix).

File: datahub/search/company/models.py [hunk 10]
```diff
@@ -131,6 +133,14 @@ class Company(BaseESModel):
...
         ),
         'latest_interaction_date': lambda obj: obj.latest_interaction_date,
+        'uk_address_postcode': dict_utils.conditional_attrgetter_with_default(
+            'uk_based',
+            'address_postcode',
+        ),
+        'uk_registered_address_postcode': dict_utils.conditional_attrgetter_with_default(
+            'uk_based',
+            'registered_address_postcode',
+        ),
```
Comment (reupen) [inline datahub/search/company/models.py]: I think lambdas would be sufficient and easier to follow than the generic function `conditional_attrgetter_with_default()` (e.g. `lambda obj: obj.address_postcode if obj.uk_based else None`).

File: datahub/search/company/test/test_entity_search_views.py [hunk 11]
```diff
@@ -372,12 +414,65 @@ def test_filters(self, setup_data, filters, expected_companies):
...
         response_data = response.json()
         assert response_data['count'] == len(expected_companies)
-
         assert [
             result['name']
             for result in response_data['results']
         ] == expected_companies
 
+    @pytest.mark.parametrize(
+        'search_term,expected_companies',
```
Comment (reupen) [inline datahub/search/company/test/test_entity_search_views.py]: Do we have a case for a lowercase `search_term`? Also, one with a full postcode?

File: datahub/search/elasticsearch.py [hunk 12]
```diff
@@ -34,6 +34,63 @@
...
+
+        # Match postcode district (with sub-district code ignored)
+        # This is so `wc1` query would match `wc1ab` and `wc1a1ab`, but not `wc111ab`
+        # Area + one or two digits
+        f'^(({POSTCODE_AREA_REGEX}[0-9]){POSTCODE_SECTOR_REGEX}{POSTCODE_UNIT_REGEX}|'
+        f'({POSTCODE_AREA_REGEX}[0-9]{{2}}){POSTCODE_SECTOR_REGEX}{POSTCODE_UNIT_REGEX}|'
+        f'({POSTCODE_AREA_REGEX}[0-9])[a-z]?{POSTCODE_SECTOR_REGEX}{POSTCODE_UNIT_REGEX})',
```
Comment (reupen) [inline datahub/search/elasticsearch.py]: Ah – actually I meant to put the entire regexes in named constants, but I think this is also a good idea. Maybe we could drop either `_REGEX` or `POSTCODE_` from these ones to make the names less repetitive?

File: changelog/elasticsearch-postcode-improvement.internal.md [hunk 13]
```diff
@@ -0,0 +1,4 @@
+`Elasticsearch`: The search of UK postcodes has been improved so that any entity having a field storing UK postcode 
+should use new `postcode_analyzer` and `postcode_search_analyzer` analyzers. It is now possible to perform partial 
+search for area, area and district, area and sub-district and also area, district and sector and whole postcode.
```
Comment (reupen) [inline changelog/elasticsearch-postcode-improvement.internal.md]: I think we could make this more concise like elsewhere (or, alternatively, use bullet points) as it's a bit hard to parse at the moment.

File: datahub/search/elasticsearch.py (hunk 12 above)
Comment (elcct) [reply datahub/search/elasticsearch.py]: I have removed the `POSTCODE_` prefix. It is now more readable. Thank you!

File: datahub/search/company/models.py (hunk 10 above)
Comment (elcct) [reply datahub/search/company/models.py]: Great idea. That function didn't really belong in dict_utils anyway, so using lambda makes perfect sense for this one-off use case. Thank you!

File: datahub/search/elasticsearch.py (hunk 9 above)
Comment (elcct) [reply datahub/search/elasticsearch.py]: That's true! Thank you!

File: changelog/elasticsearch-postcode-improvement.internal.md (hunk 13 above)
Comment (elcct) [reply changelog/elasticsearch-postcode-improvement.internal.md]: I reworded it a bit, I hope it makes more sense now.

File: datahub/search/company/test/test_entity_search_views.py (hunk 11 above)
Comment (elcct) [reply datahub/search/company/test/test_entity_search_views.py]: I added few more test cases. Thank you!
